Add optional MQTT Authentication

JumpMaster opened the pull request:
Hi,

I have added an optional MQTT Username and Password to allow for authenticated connections to an MQTT broker.

Thanks,
Kev

JumpMaster committed:
Add option MQTT Authentication

erikced commented on qt-ozwdaemon/mqttpublisher.cpp:
```diff
@@ -16,7 +16,11 @@ mqttpublisher::mqttpublisher(QSettings *settings, QObject *parent) : QObject(par
+    this->m_client->setClientId("qt-openzwave");
+    if (settings->contains("MQTTUsername") && settings->contains("MQTTPassword")) {
+        this->m_client->setUsername(settings->value("MQTTUsername", "").toString());
+	this->m_client->setPassword(settings->value("MQTTPassword", "").toString());
```
indentation?

JumpMaster replied:
This is something that frustrates me.  I used VS Code and copied the indentation from another line.  Why does it appear different on GitHub! :)

erikced replied:
Tabs (L22) vs spaces (rest of file) all over again 🙂

JumpMaster replied:
Thanks.  Fixed :)

erikced reviewed (commented):
Not a reviewer, but since I was just about to open a virtually identical PR to fix #15 ...

erikced commented on Docker/Dockerfile:
```diff
@@ -44,4 +44,4 @@ ENV BP_DB_PATH="/opt/ozw/config/crashes/"
 EXPOSE 1983
 VOLUME ["/opt/ozw/config/"]
-ENTRYPOINT /usr/local/bin/ozwdaemon -s $USB_PATH -c /opt/ozw/config/ -u /opt/ozw/config/ --mqtt-server $MQTT_SERVER --mqtt-port $MQTT_PORT --stop-on-failure --mqtt-instance $OZW_INSTANCE
+ENTRYPOINT /usr/local/bin/ozwdaemon -s $USB_PATH -c /opt/ozw/config/ -u /opt/ozw/config/ --mqtt-server $MQTT_SERVER --mqtt-port $MQTT_PORT --mqtt-username $MQTT_USERNAME --mqtt-password $MQTT_PASSWORD --stop-on-failure --mqtt-instance $OZW_INSTANCE
```
Ironically, I failed to include my only real comment: Shouldn't `$MQTT_xxx` be double quoted to avoid shell-related issues (whitespace, perhaps others as well).

Fishwaldo replied:
Actually we need a “startup script”. Right now if You don’t need any Auth, it still passes the command to ozwdaemon. Need to test if the environment variable is present and then optionally set the command line. I’d like to do that for the other options as well.

JumpMaster replied:
I did test this which is why I didn't put in defaults for the username and password.  If it's left blank it isn't picked up by the command line parser and is not added to settings.  A startup script is a cleaner approach though.

JumpMaster committed:
Fixed indentation

Fishwaldo reviewed (commented):
So I'm not a huge fan of putting passwords on command lines or config files when possible, so I'm about to push a change to your branch that turns the mqtt-password into either a enviroment file or docker secret. 

Also added a startup script and some Documentation Updates. When pushed, if you can do a quick test and confirm "works for you" that would be awesome. :)

Fishwaldo reviewed (commented):
So I'm not a huge fan of putting passwords on command lines or config files when possible, so I'm about to push a change to your branch that turns the mqtt-password into either a enviroment file or docker secret. 

Also added a startup script and some Documentation Updates. When pushed, if you can do a quick test and confirm "works for you" that would be awesome. :)

Fishwaldo committed:
Add Startup File and Change MQTT Password to Enviroment or Docker Secret

erikced commented on Docker/startozwdaemon.sh:
```diff
@@ -0,0 +1,57 @@
+	CMD="-s $USB_PATH"
+fi
+
+if [ ! -z "$MQTT_SERVER" ]
```
`MQTT_SERVER` is still defined in the `Dockerfile`, check is redundant? (applies to others as well)

Fishwaldo replied:
I don't know how other Docker Images that use this as a base image would be affected by this, hence errr on the side of Caution

erikced commented on Docker/startozwdaemon.sh:
```diff
@@ -0,0 +1,57 @@
+
+if [ ! -z "$MQTT_SERVER" ]
+then
+	CMD+=" --mqtt-server $MQTT_SERVER"
```
This does not handle whitespace very well, would be easy to fix by collecting to an array instead, e.g.
```sh
CMD=()
...
CMD+=(--mqtt-server "$MQTT_SERVER")
CMD+=(--mqtt-tls)
/usr/local/bin/ozwdaemon "${CMD[@]}"

Fishwaldo replied:
1) it appear the QT Command Line Parser doesn't handle quotes (") in arguments (it passes the quote through)
2) using a array - I'm ok.

erikced commented on Docker/startozwdaemon.sh:
```diff
@@ -0,0 +1,57 @@
+if [ ! -z "$MQTT_SERVER" ]
+then
+	CMD+=" --mqtt-server $MQTT_SERVER"
+	if [ ! -z "$MQTT_PORT" ]
```
Wouldn't it make sense to be able to change `MQTT_PORT` without changing `MQTT_SERVER`?

Fishwaldo replied:
MQTT_SERVER defaults to localhost anyway... so you would have to change the server regardless.

erikced commented on Docker/startozwdaemon.sh:
```diff
@@ -0,0 +1,57 @@
+	fi
+fi
+	
+if [ -z "$STOP_ON_FAILURE" ]
+then
+	CMD+=" --stop-on-failure"
+else 
+	if [[ $STOP_ON_FAILURE != "false" ]]
+	then
+		CMD+=" --stop-on-failure"
+	fi
+fi
+
+if [ ! -z "$MQTT_TLS" ]
+then
+	if [[ $MQTT_TLS == "true" ]]
+	then
+		CMD+=" --mqtt-tls"
+	fi
+fi
```
For `STOP_ON_FAILURE`, all values are truthy except for `false`, for `MQTT_TLS` all values are falsey except for `true`, could be confusing?

Fishwaldo replied:
its the "default" that switches them around. 
I think stop-on-failure should be enabled for the container, where as mqtt-tls is optional... 

happy to handle the logic differently though

erikced commented on Docker/startozwdaemon.sh:
```diff
@@ -0,0 +1,57 @@
+
+
+echo "Executing: /usr/local/bin/ozwdaemon $CMD"
+/usr/local/bin/ozwdaemon $CMD
```
1. Consider using `exec /usr/local/bin/ozwdaemon $CMD` to drop the shell, together with `exec` form in the dockerfile, this would allow signals to be passed on to `ozwdaemon` instead of getting caught by the shell and not forwarded.
2. Pass on any argument passed to the script, `exec /usr/local/bin/ozwdaemon $CMD "$@"` as that would allow users add custom options to the `ozwdaemon` call. (e.g. `docker run openzwave/ozwdaemon:latest --help`) if exec form is used in the dockerfile.

Fishwaldo replied:
ok. point 2 is good as well.

erikced commented on Docker/startozwdaemon.sh:
```diff
@@ -0,0 +1,57 @@
+fi
+
+
+if [ -z "$OZW_INSTANCE" ] 
```
An shorter way to set defaults would be to use parameter expansion to assign a default when you need it, e.g. `CMD+=" --mqtt-instance ${OZW_INSTANCE:=1}"`

erikced reviewed (commented):
A lot of minor comments, I can fix it up separate a PR later if you wish(/think they're useful). Setting `-u OZW_USER_DIR` and `-c $OZW_CONFIG_DIR`  appears to have been lost somewhere however?
I tested this it works for me both with `$MQTT_PASSWORD` set and `/run/secrets/MQTT_PASSWORD` mounted with a file containing the password to emulate docker swarm secrets.

Fishwaldo reviewed (commented):
If you want to build upon this, I can pull this into a branch and you can work on top of that branch with the changes.

erikced commented:
> If you want to build upon this, I can pull this into a branch and you can work on top of that branch with the changes.

I would be happy to create a patch.

JumpMaster commented:
> So I'm not a huge fan of putting passwords on command lines or config files when possible, so I'm about to push a change to your branch that turns the mqtt-password into either a enviroment file or docker secret.
> 
> Also added a startup script and some Documentation Updates. When pushed, if you can do a quick test and confirm "works for you" that would be awesome. :)

Just testing this now but I'm getting a build error.  Tried straight from master and I receive the same error:

```
The following packages have unmet dependencies:
 qtbase5-private-dev : Depends: libinput-dev but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```


EDIT
I added the below at line 15 of the Dockerfile and the build ran successfully.  Connected to mqtt as expected :)

```
RUN apt-get -y install libinput-dev libxkbcommon-dev qtbase5-dev
```

Fishwaldo commented:
@erikced 
Ok - I've pulled this into the JumpMaster-Master branch here. Please open a PR against that branch, then I'll merge to master

Fishwaldo commented:
Closing this as its merged into the JumpMaster-master branch already.
